File: frontend/src/api/control-boleto.service.ts
// Servicio del validador de boletos de la feria (escaneres dedicados).
// El tiempo real (control + monitoreo) va por WebSocket: ver ws-boletos.ts.
import http from '@/api/http'
import type {
  BoletoDentroDto,
  RegistroSalidaDetalleDto,
  RegistroSalidaDto,
  ReporteIngresosFeriaDto,
  ResumenBoletosDto,
  TipoBoleto,
  ValidacionBoletoDto,
} from '@/types/boleto.type'
import type { DiaFeria } from '@/types/boleto.type'
import type { ResultadoRegularizacionAccesoDto } from '@/types/control.type'
import type { TipoMovimiento } from '@/types/control.type'

/**
 * Valida el codigo del boleto con el escaner dedicado (tipoMovimiento),
 * DENTRO del tipo de boleto del puesto (tipoBoleto).
 * Lanza una excepcion axios si el boleto no existe en ese tipo (404) o el
 * movimiento no coincide con el estado (409, cuerpo = ValidacionBoletoDto).
 */
export async function validarBoleto(
  codigo: string,
  tipoMovimiento: TipoMovimiento,
  tipoBoleto: TipoBoleto,
): Promise<ValidacionBoletoDto> {
  const res = await http.post<ValidacionBoletoDto>(
    '/control/boletos/validar',
    { codigo, tipoMovimiento, tipoBoleto },
    { timeout: 15000 },
  )
  return res.data
}

/** Boletos que estan actualmente dentro del recinto. */
export async function boletosDentro(signal?: AbortSignal): Promise<BoletoDentroDto[]> {
  const res = await http.get<BoletoDentroDto[]>('/control/boletos/dentro', { signal, timeout: 8000 })
  return res.data
}

/** Foto del estado actual (dentro, total, ingresos y salidas). */
export async function resumenBoletos(signal?: AbortSignal): Promise<ResumenBoletosDto> {
  const res = await http.get<ResumenBoletosDto>('/control/boletos/resumen', { signal, timeout: 8000 })
  return res.data
}

/**
 * Regulariza un ingreso de feria/parqueo: registra una ENTRADA con la fecha
 * del día pedido (solo administrador). Si el boleto ya tiene ENTRADA ese día,
 * el backend responde 400 ("ya tenía registro").
 */
export async function regularizarIngresoBoleto(
  codigo: string,
  tipoBoleto: TipoBoleto,
  dia: DiaFeria,
): Promise<ResultadoRegularizacionAccesoDto> {
  const res = await http.post<ResultadoRegularizacionAccesoDto>(
    '/control/boletos/regularizar-ingreso',
    { codigo, tipoBoleto, dia },
  )
  return res.data
}

/**
 * Ingresos (solo ENTRADAS) de los 3 días de la feria, separados en
 * FERIA y PARQUEO. Base del apartado "Reportes".
 */
export async function reporteIngresosFeria(signal?: AbortSignal): Promise<ReporteIngresosFeriaDto> {
  const res = await http.get<ReporteIngresosFeriaDto>('/control/boletos/reporte-ingresos', { signal, timeout: 8000 })
  return res.data
}

/**
 * Registra los datos del visitante que dijo que va a volver.
 * Se llama DESPUÉS de la salida: la salida ya quedó registrada y este paso no
 * puede hacerla fallar.
 */
export function registrarSalida(dto: RegistroSalidaDto) {
  return http
    .post<RegistroSalidaDetalleDto>('/control/boletos/registro-salida', dto, { timeout: 20000 })
    .then((r) => r.data)
}

/**
 * Descarga la foto guardada de un registro de salida.
 * Viene de una carpeta del servidor, no de la base, y el endpoint exige token:
 * por eso se pide como blob y se arma un objectURL, igual que los PNG/PDF de los
 * tickets. Devuelve null si ese registro no tiene foto.
 *
 * Ojo: quien la use tiene que hacer `URL.revokeObjectURL` al descartarla.
 */
export function fotoDeRegistro(idRegistro: number): Promise<string | null> {
  return http
    .get('/control/boletos/registro-salida/foto', {
      params: { idRegistro },
      responseType: 'blob',
    })
    .then((r) => URL.createObjectURL(r.data as Blob))
    .catch(() => null)
}

File: src/main/java/com/uap/control_tickets/services/interfaces/ControlBoletoService.java
package com.uap.control_tickets.services.interfaces;

import com.uap.control_tickets.dto.control.BoletoDentroDto;
import com.uap.control_tickets.dto.control.ReporteIngresosFeriaDto;
import com.uap.control_tickets.dto.control.ResumenBoletosDto;
import com.uap.control_tickets.dto.control.ResultadoRegularizacionAccesoDto;
import com.uap.control_tickets.dto.control.ValidacionBoletoDto;
import com.uap.control_tickets.enums.DiaFeria;
import com.uap.control_tickets.enums.TipoAcceso;
import com.uap.control_tickets.enums.TipoBoleto;

import java.util.List;

/**
 * Validador de boletos: escaneo/tipeo del código con escáner dedicado
 * de ENTRADA o SALIDA, DENTRO de un tipo de boleto (FERIA o PARQUEO). Registra
 * el movimiento y rechaza intentos duplicados (entrar estando dentro o salir
 * estando fuera). Mismo patrón que {@link ControlService}, sin la parte de
 * matrícula (los boletos son anónimos).
 */
public interface ControlBoletoService {

    /**
     * Valida el código para el escáner indicado, dentro del tipo indicado.
     *
     * Reglas:
     *  - Código inexistente EN ESE TIPO → 404 (no válido). El mismo código
     *    puede existir en el otro tipo y es otro boleto.
     *  - ENTRADA estando dentro → 409 con motivo YA_DENTRO.
     *  - SALIDA estando fuera → 409 con motivo YA_FUERA.
     * En los casos bloqueados no se registra el movimiento ni se toca la BD.
     */
    ValidacionBoletoDto validar(String codigo, TipoAcceso tipoMovimiento, TipoBoleto tipoBoleto);

    /** Boletos actualmente dentro del recinto. */
    List<BoletoDentroDto> boletosDentro();

    /** Foto del estado actual, para pintar el monitoreo al abrir la pantalla. */
    ResumenBoletosDto resumen();

    /**
     * Regulariza UN ingreso de feria/parqueo: registra una ENTRADA con la
     * fecha del día pedido (solo ADMINISTRADOR). Si el boleto ya tiene una
     * ENTRADA ese día, se rechaza ("ya tenía registro"). El "dentro" solo se
     * toca si el día pedido es hoy.
     */
    ResultadoRegularizacionAccesoDto regularizarIngreso(
            String codigo, TipoBoleto tipoBoleto, DiaFeria dia);

    /**
     * Ingresos (solo ENTRADAS) de los 3 días de la feria, separados en
     * FERIA y PARQUEO. Base del apartado "Reportes" y de su PDF.
     */
    ReporteIngresosFeriaDto reporteIngresosPorDia();

    /**
     * Cierra la jornada: deja a TODOS (tickets y boletos) como fuera del recinto.
     * El sistema ya limpia solo el 'dentro' que quedo de un dia anterior cuando la
     * persona vuelve a escanear; esto es el boton manual para dejar los contadores
     * en cero al terminar el dia, sin esperar al proximo escaneo.
     *
     * @return cuantos tickets y boletos se pusieron en fuera.
     */
    int cerrarJornada();

    /**
     * Guarda los datos que dio el visitante al salir diciendo que va a volver.
     * Los tres campos son opcionales; sin ninguno queda `sinDatos=true`.
     */
    com.uap.control_tickets.dto.control.RegistroSalidaDetalleDto registrarSalida(
            com.uap.control_tickets.dto.control.RegistroSalidaDto dto);

    /**
     * Bytes (JPEG) de la foto de un registro de salida, o null si no tiene.
     * La imagen se guarda en una carpeta del disco, no en la BD.
     */
    byte[] fotoDeRegistro(Long idRegistro);
}
